### PR TITLE
TextStyle level leadingDistribution

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
The framework failure seems to be caused by some `toString` changes.
I'll check with Christian and see if this produces the correct graphics. 

* https://github.com/flutter/flutter/issues/72521

## Screencast 
* There're 2 different `TextSpan`s rendered side by side: "firstTextRunljx" and "secondTextRunljx".
* Their styles are controlled by the 2nd and 3rd groups of controls. The sliders control the height multipliers, ranging from 0.0x - 20.0x.
* The blue box rendered behind the text is the space the `TextSpan`s occupied as seen by the framework

https://user-images.githubusercontent.com/31859944/107083956-433be780-67ab-11eb-99c0-bdc462f730f5.mp4

## TODO (after this PR)
- Skia & canvaskit 
- CSS implementation (add AD Scaling?)
- Framework & documentation

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.
- [ ] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

File: third_party/txt/src/txt/paragraph_txt.cc [hunk 1]
```diff
@@ -1186,79 +1193,92 @@ void ParagraphTxt::UpdateLineMetrics(const SkFontMetrics& metrics,
...
+    // https://glyphsapp.com/tutorials/vertical-metrics
+    //
+    // Doing this ascent/descent normalization to the EM Square allows
+    // a sane, consistent, and reasonable "blob_height" to be specified,
+    // though it breaks with what is done by any of the platforms above.
+    const bool shouldNormalizeFont = style.has_height_override;
+    const double font_height =
+        shouldNormalizeFont ? style.font_size : metrics_font_height;
+
+    const size_t text_height_behavior =
```
Comment (LongCatIsLooong) [inline third_party/txt/src/txt/paragraph_txt.cc:1253]: Defaults to `ParagraphStyle.textHeightBehavior` if `TextStyle.textHeightBehavior` is not specified. 

In the framework `TextPainter` does not know about `TextStyle.height` but currently has a `textHeightBehavior` field. With this implementation `TextPainter.textHeightBehavior` has the lowest precedence in the `textHeightBehavior` defaulting rules.

https://github.com/flutter/flutter/blob/b926c7b696eb1a5f601ef74cd191b3b6b2b22729/packages/flutter/lib/src/painting/text_painter.dart#L411-L437

File: lib/ui/text.dart [hunk 2]
```diff
@@ -432,14 +432,43 @@ enum TextDecorationStyle {
   wavy
 }
 
+/// {@macro dart.ui.leadingDistribution}
+enum LeadingDistribution {
+  /// Distributes the ["leading"](https://www.w3.org/TR/CSS2/visudet.html#leading)
```
Comment (Hixie) [inline lib/ui/text.dart:437]: let's link to https://en.wikipedia.org/wiki/Leading rather than the CSS spec though

File: lib/ui/text.dart [hunk 3]
```diff
@@ -432,14 +432,43 @@ enum TextDecorationStyle {
...
+  /// space of the text proportionally above and below the text, to the
+  /// font's ascent/discent ratio.
+  ///
+  /// {@template dart.ui.leading}
+  /// The leading space of a text run is defined as
```
Comment (Hixie) [inline lib/ui/text.dart:442]: "leading" is a noun, not an adjective. (It's pronounced like the metal Pb, as opposed to like the opposite of "following")

Comment (Hixie) [reply lib/ui/text.dart:442]: (which is to say, it's "the leading", not "the leading space".)

File: lib/ui/text.dart [hunk 4]
```diff
@@ -810,6 +869,7 @@ class TextStyle {
              'letterSpacing: ${      _encoded[0] & 0x00800 == 0x00800  ? "${_letterSpacing}x"                    : "unspecified"}, '
              'wordSpacing: ${        _encoded[0] & 0x01000 == 0x01000  ? "${_wordSpacing}x"                      : "unspecified"}, '
              'height: ${             _encoded[0] & 0x02000 == 0x02000  ? "${_height}x"                           : "unspecified"}, '
+             'textHeightBehavior: ${ _encoded[0] & 0x0100 == 0x0100    ? "${TextHeightBehavior.fromEncoded(_encoded[8])}x"  : "unspecified"}, '
```
Comment (Hixie) [inline lib/ui/text.dart:872]: i suspect the "x" here is a copy/paste typo?

File: third_party/txt/src/txt/text_style.h [hunk 5]
```diff
@@ -31,6 +31,26 @@
...
+// metric provided ascent/descent and ParagraphStyle.height will not take
+// effect.
+//
+// The default behavior is kAll where height adjustments are enabled for all
+// lines.
+//
+// Multiple behaviors can be applied at once with a bitwise | operator. For
+// example, disabling first ascent and last descent can achieved with:
+//
+//   (kDisableFirstAscent | kDisableLastDescent).
```
Comment (Hixie) [inline third_party/txt/src/txt/text_style.h:45]: we should update the docs to mention half-leading too

**Comment by Hixie** [comment]
glancing through this it seems reasonable. (didn't do a detailed review)

File: lib/ui/text.dart (hunk 2 above)
Comment (LongCatIsLooong) [reply lib/ui/text.dart:437]: Would it be better to avoid using the term altogether then? The Wikipedia page says "the exact definition varies", and "In electronic typesetting, "leading" often instead means the distance from one baseline to the next. ".

File: third_party/txt/src/txt/paragraph_txt.cc [hunk 6]
```diff
@@ -1186,79 +1197,98 @@ void ParagraphTxt::UpdateLineMetrics(const SkFontMetrics& metrics,
...
+    const bool shouldNormalizeFont =
+        style.has_height_override &&
+        !(text_height_behavior & TextHeightBehavior::kHalfLeading);
+    const double font_height =
+        shouldNormalizeFont ? style.font_size : metrics_font_height;
+
+    const double leading =
+        text_height_behavior & TextHeightBehavior::kHalfLeading
+            ? blob_height - font_height
+        : style.has_height_override ? 0.0
```
Comment (johnsonmh) [inline third_party/txt/src/txt/paragraph_txt.cc:1268]: I can't quite figure out this part of the ternary. This `0.0` will be hit when: "if not half leading, and style has a height override, then leading == 0"? Is that because we don't consider there to be any "leading" with AD scaling, instead of adding leading we just stretch the ascent and descent?

Comment (johnsonmh) [reply third_party/txt/src/txt/paragraph_txt.cc:1268]: Side note and totally personal preference so feel free to ignore: it takes me a few extra seconds to parse the nested ternary, I personally would find something like this easier to read:

```cpp
if (halfLeading) {
  leading = blob_height - font_height;
else {
  leading = style.has_height_override ? 0.0 : metrics.fLeading;
}

File: third_party/txt/src/txt/paragraph_txt.cc [hunk 7]
```diff
@@ -1186,79 +1197,98 @@ void ParagraphTxt::UpdateLineMetrics(const SkFontMetrics& metrics,
...
+        text_height_behavior & TextHeightBehavior::kDisableLastDescent;
+
+    // Proportionally distribute the remaining vertical space above and below
+    // the glyph blob's baseline, per the font's ascent/discent ratio.
+    const double modifiedAscent =
+        -metrics.fAscent / metrics_font_height * available_vspace +
+        half_leading;
+    double ascent = disableAscent ? -metrics.fAscent : modifiedAscent;
+    double descent = disableDescent ? metrics.fDescent
+                                    //: blob_height - leading - modifiedAscent;
```
Comment (johnsonmh) [inline third_party/txt/src/txt/paragraph_txt.cc:1288]: This comment should probably be removed.

File: third_party/txt/src/txt/paragraph_txt.cc [hunk 8]
```diff
@@ -1186,79 +1197,98 @@ void ParagraphTxt::UpdateLineMetrics(const SkFontMetrics& metrics,
...
+                                    : metrics.fDescent / metrics_font_height *
```
Comment (johnsonmh) [inline third_party/txt/src/txt/paragraph_txt.cc:1289]: For consistency, consider putting this in a `double` called `modifiedDescent`.

File: third_party/txt/src/txt/paragraph_txt.cc [hunk 9]
```diff
@@ -671,7 +682,7 @@ void ParagraphTxt::Layout(double width) {
   glyph_lines_.clear();
   code_unit_runs_.clear();
   inline_placeholder_code_unit_runs_.clear();
-  max_right_ = FLT_MIN;
+  max_right_ = -FLT_MAX;
```
Comment (johnsonmh) [inline third_party/txt/src/txt/paragraph_txt.cc:685]: I'm not familiar with what this does, is this change related/required by the current change? Will it have any side effects, or is `FLT_MIN` always the `-` of `FLT_MAX`?

File: third_party/txt/src/txt/text_style.cc [hunk 10]
```diff
@@ -48,6 +48,12 @@ bool TextStyle::equals(const TextStyle& other) const {
     return false;
   if (has_height_override != other.has_height_override)
     return false;
+  if (has_text_height_behavior_override !=
+      other.has_text_height_behavior_override)
+    return false;
+  if (has_text_height_behavior_override &&
```
Comment (johnsonmh) [inline third_party/txt/src/txt/text_style.cc:54]: why not just `text_height_behavior != other.text_height_behavior`?

**Review by johnsonmh — APPROVED** [review]
Looks great!

**Comment by chinmaygarde** [comment]
@Hixie Are you still planning on the detailed review?

**Comment by chinmaygarde** [comment]
Looks like the presubmit failure is real. The test needs to be updated.

File: third_party/txt/src/txt/paragraph_txt.cc (hunk 9 above)
Comment (LongCatIsLooong) [reply third_party/txt/src/txt/paragraph_txt.cc:685]: According to the documentation `FLT_MIN` is the smallest positive `double` (that's not `0`). I had to change this for some strut related stuff so descent can go negative. Some of other occurrences may not be directly related to this change but they all seem to make more sense with `-FLT_MAX`.

File: lib/ui/text/paragraph_builder.cc [hunk 11]
```diff
@@ -118,14 +120,16 @@ const int sFontStyleIndex = 1;
...
+const int sTextHeightBehaviorIndex = 5;
+const int sLeadingIndex = 6;
+const int sForceStrutHeightIndex = 7;
 
 const int sFontWeightMask = 1 << sFontWeightIndex;
 const int sFontStyleMask = 1 << sFontStyleIndex;
 const int sFontFamilyMask = 1 << sFontFamilyIndex;
 const int sFontSizeMask = 1 << sFontSizeIndex;
 const int sHeightMask = 1 << sHeightIndex;
+const int stextHeightBehaviorMask = 1 << sTextHeightBehaviorIndex;
```
Comment (jason-simmons) [inline lib/ui/text/paragraph_builder.cc:132]: Capitalize this as `sTextHeightBehaviorMask`

File: third_party/txt/src/txt/paragraph_txt.cc (hunk 9 above)
Comment (jason-simmons) [reply third_party/txt/src/txt/paragraph_txt.cc:685]: Yes - I think all of these initializers should have been `-FLT_MAX`.  They are all used in `std::max(current_value, new_value)` expressions, and the goal is to ensure that any `new_value` is greater than the initialized `current_value`.

Another option would be using `std::numeric_limits<double>::lowest()`

Comment (LongCatIsLooong) [reply third_party/txt/src/txt/paragraph_txt.cc]: Switched to std::numeric_limits<T> as they all seem to be `double`s.

**Comment by LongCatIsLooong** [comment]
@johnsonmh is it also desirable to configure `disableFirstAscent` from within a `TextTheme`, or are there any other benefits of introducing a `TextStyle` level `disableFirstAscent`? `disableFirstAscent` is likely not gonna work if any `TextStyle`s used by the first line decides to not disable its first ascent (which kinda depends on where do we break the first line).

**Comment by chinmaygarde** [comment]
@johnsonmh: Any updates to @LongCatIsLooong s query?
@LongCatIsLooong: If no progress on this PR is likely, can we close it and followup later please?

**Comment by johnsonmh** [comment]
Thanks for the ping @chinmaygarde, apologies for the delay @LongCatIsLooong.

To answer your question

> is it also desirable to configure disableFirstAscent from within a TextTheme, or are there any other benefits of introducing a TextStyle level disableFirstAscent?

I would say this is not nearly as important or even desirable as having half leading available on the text theme level somehow. `disableFirstAscent` is mostly used to solve the issues that arise from using proportional leading, so I suspect there will not be demand for it as long as users can opt into half-leading, which solves the same issues.

**Comment by LongCatIsLooong** [comment]
Updated the PR to:

- add `LeadingDistribution` to `TextHeightBehavior`, `TextStyle` and `StrutStyle` (instead of adding it to `TextHeightBehavior` and adding `TextHeightBehavior` to `TextStyle` and `StrutStyle`). How it works largely remain the same.  
- move some `ui.TextStyle.toString` tests from framework to `text_test.dart`

**Comment by LongCatIsLooong** [comment]
Ready for review.
@johnsonmh @Hixie @jason-simmons there's an API change since you last reviewed the PR. Could you take a look?

File: lib/ui/text.dart [hunk 12]
```diff
@@ -1565,14 +1565,46 @@ enum TextDecorationStyle {
   wavy
 }
 
+/// {@macro dart.ui.leadingDistribution}
+enum LeadingDistribution {
```
Comment (GaryQian) [inline lib/ui/text.dart:1569]: Optional, but it seems that most text related enums/classes are prefixed with `Text`.

I don't have a particularly strong conviction either way, but It would also be reasonable to call this `TextLeadingDistribution` for consistency. That being said, as far as I know, the concept of leading is unique to text layout, so this naming should be sufficient (currently) to identify the feature uniquely.

File: lib/ui/text.dart [hunk 13]
```diff
@@ -2278,6 +2352,11 @@ class StrutStyle {
   /// * `leading`: The minimum amount of leading between lines as a multiple of
```
Comment (GaryQian) [inline lib/ui/text.dart:2352]: Probably want to improve documentation on this property as well to better distinguish it, since there are now two similarly named properties right next to each other.

File: third_party/txt/src/txt/paragraph_style.h [hunk 14]
```diff
@@ -58,6 +66,7 @@ enum TextHeightBehavior {
   kDisableFirstAscent = 0x1,
   kDisableLastDescent = 0x2,
   kDisableAll = 0x1 | 0x2,
+  kHalfLeading = 0x1 << 2,
```
Comment (GaryQian) [inline third_party/txt/src/txt/paragraph_style.h:69]: I would prefer consistent naming in the engine and the framework. This was called `even` in framework. Two names for the same property makes this much more confusing to read and maintain.

File: lib/web_ui/lib/src/engine/text/paragraph.dart [hunk 15]
```diff
@@ -1211,6 +1211,8 @@ class EngineStrutStyle implements ui.StrutStyle {
     List<String>? fontFamilyFallback,
     double? fontSize,
     double? height,
+    //TODO(LongCatIsLooong): implement leadingDistribution.
```
Comment (johnsonmh) [inline lib/web_ui/lib/src/engine/text/paragraph.dart:1214]: is this TODO meant to be here? It looks like this is implemented?

**Review by johnsonmh — APPROVED** [review]
LGTM

File: lib/web_ui/lib/src/engine/text/paragraph.dart (hunk 15 above)
Comment (LongCatIsLooong) [reply lib/web_ui/lib/src/engine/text/paragraph.dart:1214]: by that I meant currently setting `leadingDistribution` doesn't do anything on web and it needs to be implemented.
